devito/propagator.py

```diff
@@ -30,10 +33,20 @@ def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0):
         # Start with the assumption that the propagator needs to save the field in memory at every time step
         self._save = True
         # This might be changed later when parameters are being set
-
+        self.profile = os.environ.get(self._ENV_VAR_PROFILE) == "1"
         # Which function parameters need special (non-save) time stepping and which don't
         self.save_vars = {}
         self.fd = FunctionDescriptor(name)
+        self.pre_loop = []
+        self.post_loop = []
+        if self.profile:
+            self.add_local_var("time", "double")
+            self.pre_loop.append(cgen.Statement("struct timeval start, end"))
+            self.pre_loop.append(cgen.Assign("time", 0))
+            self.post_loop.append(cgen.PrintStatement("time: %f\n", "time"))
+            self.time_loop_stencils_b.append(cgen.Statement("gettimeofday(&start, NULL)"))
+            self.time_loop_stencils_a.append(cgen.Statement("gettimeofday(&end, NULL)"))
+            self.time_loop_stencils_a.append(cgen.Statement("time += ((end.tv_sec  - start.tv_sec) * 1000000u + end.tv_usec - start.tv_usec) / 1.e6"))
```

Comment:
Ok - you are touching an important issue here which I thinks needs pulling out into its own issue and branch. Currently the way the boundary condition is implemented is as a sponge layer implemented with the matrix 'damp'. The problem with this implementation is that it is zero for most of the domain (wasted flops) and it increases memory traffic by about a third in the acoustic case... So we really need to reimplement this by splitting the space loop between the boundary region and the internal region. @navjotk, @mloubout - can you add this to your todo list please?

BTW - I don't think this should hold up this pr.